From 2d1341731f99c133563032539cad9ec59c99c2b5 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:02:04 -0400
Subject: [infiniband] mlx4: Avoid executing gid task when device is being removed

Message-id: <0780f7589268f725e0bd70eff519cd5a454fa711.1410425016.git.avadai@redhat.com>
Patchwork-id: 93724
O-Subject: [Patch RHEL7.1 V1 188/197] IB/mlx4: Avoid executing gid task when device is being removed
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Moni Shoua <monis@mellanox.com>

When device is being removed (e.g during VPI port link type change from ETH
to IB), tasks for gid table changes should not be executed.

Flush the current queue of tasks and block further tasks from entering the queue.

Signed-off-by: Moni Shoua <monis@mellanox.com>
Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>

Upstream: 4bf9715f184969dc703bde7be94919995024a6a9
Signed-off-by: Amir Vadai <avadai@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/mlx4/main.c b/drivers/infiniband/hw/mlx4/main.c
index 45a98a3..3242fbb 100644
--- a/drivers/infiniband/hw/mlx4/main.c
+++ b/drivers/infiniband/hw/mlx4/main.c
@@ -1387,6 +1387,9 @@ static void update_gids_task(struct work_struct *work)
  int err;
  struct mlx4_dev *dev = gw->dev->dev;
 
+ if (!gw->dev->ib_active)
+  return;
+
  mailbox = mlx4_alloc_cmd_mailbox(dev);
  if (IS_ERR(mailbox)) {
   pr_warn("update gid table failed %ld\n", PTR_ERR(mailbox));
@@ -1417,6 +1420,9 @@ static void reset_gids_task(struct work_struct *work)
  int err;
  struct mlx4_dev *dev = gw->dev->dev;
 
+ if (!gw->dev->ib_active)
+  return;
+
  mailbox = mlx4_alloc_cmd_mailbox(dev);
  if (IS_ERR(mailbox)) {
   pr_warn("reset gid table failed\n");
@@ -2366,6 +2372,9 @@ static void mlx4_ib_remove(struct mlx4_dev *dev, void *ibdev_ptr)
  struct mlx4_ib_dev *ibdev = ibdev_ptr;
  int p;
 
+ ibdev->ib_active = false;
+ flush_workqueue(wq);
+
  mlx4_ib_close_sriov(ibdev);
  mlx4_ib_mad_cleanup(ibdev);
  ib_unregister_device(&ibdev->ib_dev);
-- 
1.7.1